src/PipelineChannel.php

```diff
@@ -0,0 +1,67 @@
+<?php
+
+namespace Amp\Sync;
+
+use Amp\ByteStream\ClosableStream;
+use Amp\Cancellation;
+use Amp\Pipeline\Emitter;
+use Amp\Pipeline\Pipeline;
+
+/**
+ * Creates a channel from a Pipeline and Emitter. The Pipeline emits data to be received on the channel (data
+ * emitted on the Pipeline will be returned from calls to {@see Channel::receive()}). The Emitter will receive data
+ * that sent on the channel (data passed to {@see Channel::send()} will be passed to {@see Emitter::yield()}).
+ *
+ * @template TReceive
+ * @template TSend
+ * @template-implements Channel<TReceive, TSend>
+ */
+final class PipelineChannel implements Channel, ClosableStream
+{
+    /**
+     * @param Pipeline<TReceive> $receive
+     * @param Emitter<TSend> $send
+     */
+    public function __construct(
+        private Pipeline $receive,
+        private Emitter $send,
+    ) {
+    }
+
+    public function __destruct()
+    {
+        $this->close();
+    }
+
+    public function isClosed(): bool
+    {
+        return $this->send->isComplete();
+    }
+
+    public function close(): void
+    {
+        if (!$this->send->isComplete()) {
+            $this->send->complete();
+        }
+
+        $this->receive->dispose();
+    }
+
+    public function receive(?Cancellation $cancellation = null): mixed
+    {
+        return $this->receive->continue($cancellation);
+    }
+
+    public function send(mixed $data): void
+    {
+        if ($data === null) {
+            throw new ChannelException("Cannot send null on a channel");
```

Comment:
Can we allow this now?

Reply:
`null` is currently used to indicate that the channel has closed when returned from `Channel::receive()`.